migration/v0.43/apply/20-upgrade-opensearch.sh

```diff
@@ -0,0 +1,163 @@
+  {
+    "settings": {
+      "index.blocks.write": true
+    }
+  }
+  ')
+
+  acknowledged=$(echo "${resp}" | grep "^{" | jq -r '.acknowledged')
+  if [ "${acknowledged}" = "true" ]; then
+    log_info "- Marked '${os_dashboards_index}' as read-only"
+  else
+    log_fatal "Failed to mark index '${os_dashboards_index}' as read-only" "${resp}"
+  fi
+
+  log_info "- Cloning index '${os_dashboards_index}' to index .kibana_1"
+  resp=$(curl -sS -kL -u "${user}:${password}" -X PUT "${os_url}"/"${os_dashboards_index}"/_clone/.kibana_1)
+  acknowledged=$(echo "${resp}" | grep "^{" | jq -r '.acknowledged')
+  if [ "${acknowledged}" = "true" ]; then
+    log_info "- Successfully cloned '${os_dashboards_index}' to .kibana_1"
+  else
+    log_fatal "Failed to clone index '${os_dashboards_index}' to .kibana_1" "${resp}"
+  fi
+
+}
+
+disable_read_only_kibana_index() {
+
+  log_info "- Disabling read-only mode for .kibana_1"
+  resp=$(curl -sS -kL -u "${user}:${password}" -X PUT "${os_url}"/.kibana_1/_settings -H 'Content-Type: application/json' -d'
+  {
+    "settings": {
+      "index.blocks.write": false
+    }
+  }
+  ')
+  acknowledged=$(echo "${resp}" | grep "^{" | jq -r '.acknowledged')
+  if [ "${acknowledged}" = "true" ]; then
+    log_info "- Successfully disabled read-only mode for .kibana_1"
+  else
+    log_fatal "Failed to disable read-only mode for .kibana_1" "${resp}"
+  fi
+
+}
+
+create_kibana_alias() {
+
+  resp=$(curl -sS -kL -u "${user}:${password}" -X GET "${os_url}"/_alias/.kibana)
+  if [[ $(echo "${resp}" | jq -r 'to_entries | .[0].key') == .kibana* ]]; then
+    log_info "- Alias .kibana already exists, skipping"
+  elif [[ $(echo "${resp}" | jq -r '.status') == "404" ]]; then
+    log_info "- Creating alias .kibana"
+    resp=$(curl -sS -kL -u "${user}:${password}" -X PUT "${os_url}"/.kibana_1/_aliases/.kibana)
+    acknowledged=$(echo "${resp}" | grep "^{" | jq -r '.acknowledged')
+    if [ "${acknowledged}" = "true" ]; then
+      log_info "- Successfully created alias .kibana"
+    else
+      log_fatal "Failed to create alias .kibana" "${resp}"
+    fi
+  else
+    log_fatal "Failed to check if alias .kibana exists" "${resp}"
+  fi
+
+}
+
+delete_opensearch_dashboards_index() {
+
+  os_dashboards_index=$(curl -sS -kL -u "${user}:${password}" -X GET "${os_url}"/_alias/.opensearch_dashboards | jq -r 'to_entries | .[0].key')
+  if [[ "${os_dashboards_index}" != .opensearch_dashboards* ]]; then
+    log_info "- Skipping: Alias .opensearch_dashboards doesn't exist"
+  else
+    log_info "- Deleting index '${os_dashboards_index}'"
+    resp=$(curl -sS -kL -u "${user}:${password}" -X DELETE "${os_url}"/"${os_dashboards_index}")
+    acknowledged=$(echo "${resp}" | grep "^{" | jq -r '.acknowledged')
+    if [ "${acknowledged}" = "true" ]; then
+      log_info "- Successfully deleted index '${os_dashboards_index}'"
+    else
+      log_fatal "Failed to delete index '${os_dashboards_index}'" "${resp}"
+    fi
+  fi
+
+}
+
+run() {
```

Comment:
Is there any idea of shutting down opensearch-dashboards while performing the index/alias migration?

Reply:
I guess if dashboards crashes and restarts after the `.opensearch_dashboards` index has been deleted but before the changes are applied, it would recreate the index and it could get quite cumbersome to get rid of after upgrading.

Reply:
I would then argue for that it would be beneficial to shutdown dashboards before the index migration :)